### PR TITLE
Add caching to HeaderChain struct

### PR DESCRIPTION
**Pull request by adrianbrink**
This is the initial PR for adding caching to the HeaderChain struct.

- spelling fixes in ethcore-light
- add reference to docs for LRU cache
- add member variable ```cache``` on HeaderChain

Left todo:
- use the cache in the methods

#5398

### CODE REVIEW

File: ethcore/light/src/client/header_chain.rs [hunk 1]
```diff
@@ -408,7 +414,7 @@ impl HeaderChain {
 		}
 	}
 
-	/// Get the nth CHT root, if it's been computed.
+	/// Get the nth CHT root, if it has been computed.
```
Comment (rphmeier) [inline ethcore/light/src/client/header_chain.rs]: technically not wrong before but reads better here :)

File: ethcore/light/src/net/mod.rs [hunk 2]
```diff
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
-//! PIP Protocol Version 1 implementation.
+//! PLP Protocol Version 1 implementation.
```
Comment (rphmeier) [inline ethcore/light/src/net/mod.rs]: No, it really is PIP...

maybe I'll just change the name if it's really confusing :)

**Comment by rphmeier** [comment]
Thanks! The key method to look at is `block_header` for utilizing the cache.

File: ethcore/light/src/net/mod.rs (hunk 2 above)
Comment (adrianbrink) [inline ethcore/light/src/net/mod.rs]: But isn't it **P**arity **L**ightlient **P**rotocol? I have to say that I find it pretty confusing :-)

**Comment by adrianbrink** [comment]
I've rebased and now update the cache.

It looks a build verbose due to the fact that ```insert_block_header()``` doesn't return the inserted item and hence the extended function block in ```and_then()```.

File: ethcore/light/src/client/header_chain.rs (hunk 1 above)
Comment (adrianbrink) [inline ethcore/light/src/client/header_chain.rs]: Reversed

File: ethcore/light/src/net/mod.rs (hunk 2 above)
Comment (adrianbrink) [inline ethcore/light/src/net/mod.rs]: Also reversed.

File: ethcore/light/src/client/header_chain.rs [hunk 3]
```diff
@@ -359,11 +365,24 @@ impl HeaderChain {
...
-					warn!(target: "chain", "Failed to read from database: {}", e);
-					None
+			let mut cache = self.cache.lock();
+
+			match cache.block_header(&hash) {
+				Some(header) => Some(header),
+				None => {
+					match self.db.get(self.col, &hash) {
+						Ok(dbValue) => {
+							dbValue.map(|x| x.to_vec()).map(encoded::Header::new)
```
Comment (rphmeier) [inline ethcore/light/src/client/header_chain.rs]: last thing: in Rust we usually use `snake_case` style. So `dbValue` should be `db_value`

**Comment by rphmeier** [comment]
Okay, LGTM except for the capitalization issue

File: ethcore/light/src/client/service.rs [hunk 4]
```diff
@@ -56,6 +60,8 @@ pub struct Service {
 impl Service {
 	/// Start the service: initialize I/O workers and client itself.
 	pub fn start(config: ClientConfig, spec: &Spec, path: &Path) -> Result<Self, Error> {
+		let cache = Arc::new(Mutex::new(Cache::new(Default::default(), Duration::hours(6))));
```
Comment (rphmeier) [inline ethcore/light/src/client/service.rs]: This should be given a cache, which should be the same one given to `OnDemand` and RPC in `parity/run.rs`

**Comment by adrianbrink** [comment]
@rphmeier Rebased and fixed up parity/run.rs to now use the same Cache. However, currently we are passing the cache size down but I don't think it is used. I can try and fix that in another PR.

**Comment by rphmeier** [comment]
Thanks!

LGTM, seems like some random other changes have gotten mixed into the diff, though.
Also, the CI doesn't appear to have run...

**Comment by adrianbrink** [comment]
Anything that I should fix? Aren't those other changes from me rebasing?

**Comment by gavofyork** [comment]
will push this as a branch in our repo to get the fire

**Comment by gavofyork** [comment]
see #5428

**Comment by rphmeier** [comment]
Looks like there are some test build failures.
Can you run `test.sh` until it at builds and completes successfully? Maybe set `CARGO_INCREMENTAL=1` for faster compilation.

**Comment by adrianbrink** [comment]
Just noticed. I'll fix it now and then I'll update the PR. And also, I'll run the test script. Before I only ran the tests for the light_client crate.

**Comment by adrianbrink** [comment]
@gavofyork @rphmeier  
The tests in ```./test.sh``` are all passing now locally. Could you trigger a 2nd CI build.

Also, do any of you have a good guide on how to handle rebasing when using a forking approach to git?

**Comment by adrianbrink** [comment]
Also, is there a way to speed up the tests for the entire parity build?

I'm running them with ```CARGO_INCREMENTAL=1 ./test.sh``` on a MBP 13" and it takes around 5 min. :-)

**Comment by rphmeier** [comment]
Usually we just `git merge` against the latest master rather than rebase because it's clean.

And that's about as fast as the tests get :)

**Comment by adrianbrink** [comment]
Okay, that's good to know. I'll switch to ```git merge``` next time :-)

**Pull request merged**